stores/slabbuffer.go

```diff
@@ -0,0 +1,505 @@
+			if _, exists := mgr.incompleteBuffers[gid]; exists {
+				mgr.completeBuffers[gid] = append(mgr.completeBuffers[gid], buffer)
+				for i := range mgr.incompleteBuffers[gid] {
+					if mgr.incompleteBuffers[gid][i] == buffer {
+						mgr.incompleteBuffers[gid] = append(mgr.incompleteBuffers[gid][:i], mgr.incompleteBuffers[gid][i+1:]...)
+						break
+					}
+				}
+			}
+			mgr.mu.Unlock()
+		}
+		// Remember to update the db with the new size if necessary.
+		dbUpdates = append(dbUpdates, dbUpdate{
+			buffer:   buffer,
+			complete: complete,
+			syncSize: syncSize,
+		})
+	}
+
+	// Update size field in db. Since multiple threads might be trying to do
+	// this, the operation is associative.
+	maxOp := "GREATEST"
+	if isSQLite(mgr.s.db) {
+		maxOp = "MAX"
+	}
+	for _, update := range dbUpdates {
+		err = func() error {
+			// Make sure only one thread can update the entry for a buffer at a
+			// time.
+			update.buffer.dbMu.Lock()
+			defer update.buffer.dbMu.Unlock()
+
+			// Since the order in which threads arrive here is not deterministic
+			// there is a chance that we don't need to perform this update
+			// because a larger size was already written to the db.
+			if !update.buffer.requiresDBUpdate() {
+				return nil
+			}
+			err = mgr.s.retryTransaction(func(tx *gorm.DB) error {
+				if err := tx.Model(&dbBufferedSlab{}).
+					Where("id", update.buffer.dbID).
+					Updates(map[string]interface{}{
+						"complete": gorm.Expr("complete OR ?", update.complete),
+						"size":     gorm.Expr(maxOp+"(size, ?)", update.syncSize),
+					}).
+					Error; err != nil {
+					return fmt.Errorf("failed to update buffered slab %v in database: %v", update.buffer.dbID, err)
+				}
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+			// Update the dbSize field in the buffer to the size we just wrote to the
+			// db. This also needs to be associative.
+			update.buffer.updateDBSize(update.syncSize)
+			return nil
+		}()
+		if err != nil {
+			return nil, fmt.Errorf("failed to update size/complete in db: %w", err)
+		}
+	}
+	return slabs, nil
+}
+
+func (mgr *SlabBufferManager) FetchPartialSlab(ctx context.Context, ec object.EncryptionKey, offset, length uint32) ([]byte, error) {
+	mgr.mu.Lock()
+	buffer, exists := mgr.buffersByKey[ec.String()]
+	mgr.mu.Unlock()
+	if !exists {
+		return nil, api.ErrObjectNotFound
+	}
+
+	data := make([]byte, length)
+	_, err := buffer.file.ReadAt(data, int64(offset))
+	return data, err
+}
+
+func (mgr *SlabBufferManager) SlabBuffers() []api.SlabBuffer {
+	var sbs []api.SlabBuffer
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	convertBuffer := func(buffer *SlabBuffer, complete bool) api.SlabBuffer {
+		return api.SlabBuffer{
+			ContractSet: "", // filled in by caller
+			Complete:    complete,
+			Filename:    buffer.filename,
+			Size:        buffer.size,
+			MaxSize:     buffer.maxSize,
+			Locked:      time.Now().Before(buffer.lockedUntil),
+		}
+	}
+	for _, buffers := range mgr.completeBuffers {
```

Comment:
`nit` you could change these two `for` loops into a single one with `append(complete, incomplete...)` and use the index as the conditional

Reply:
Yeah but readability > than saving 3 lines in this case.